libs/pandautils/src/main/java/com/instructure/pandautils/features/inbox/list/InboxRepository.kt

```
@@ -0,0 +1,117 @@
+
+class InboxRepository(
+    private val inboxApi: InboxApi.InboxInterface,
+    private val coursesApi: CourseAPI.CoursesInterface,
+    private val groupsApi: GroupAPI.GroupInterface,
+    private val progressApi: ProgressAPI.ProgressInterface
+) {
+
+    suspend fun getConversations(scope: InboxApi.Scope, forceNetwork: Boolean, filter: CanvasContext?, nextPageLink: String? = null): DataResult<List<Conversation>> {
+        val params = RestParams(usePerPageQueryParam = true, isForceReadFromNetwork = forceNetwork)
+        return if (nextPageLink == null) {
+            if (filter == null) {
+                getConversationsAllCourses(scope, params)
+            } else {
+                getConversationsFiltered(scope, filter.contextId, params)
+            }
+        } else {
+            inboxApi.getNextPage(nextPageLink, params)
+        }
+    }
+
+    private suspend fun getConversationsAllCourses(scope: InboxApi.Scope, params: RestParams): DataResult<List<Conversation>> {
+        return inboxApi.getConversations(InboxApi.conversationScopeToString(scope), params)
+    }
+
+    private suspend fun getConversationsFiltered(scope: InboxApi.Scope, contextId: String, params: RestParams): DataResult<List<Conversation>> {
+        return inboxApi.getConversationsFiltered(InboxApi.conversationScopeToString(scope), contextId, params)
+    }
+
+    suspend fun getCanvasContexts(): DataResult<List<CanvasContext>> {
+        val params = RestParams(usePerPageQueryParam = true)
+
+        val coursesResult = coursesApi.getFirstPageCourses(params)
+            .depaginate { nextUrl -> coursesApi.next(nextUrl, params) }
+
+        if (coursesResult.isFail) return coursesResult
+
+        val groupsResult = groupsApi.getFirstPageGroups(params)
+            .depaginate { nextUrl -> groupsApi.getNextPageGroups(nextUrl, params) }
+
+        val courses = (coursesResult as DataResult.Success).data
+        val groups = groupsResult.dataOrNull ?: emptyList()
```

We're hiding this error, shouldn't this be dataOrThrow instead?

---

It's an edge case but we don't want to throw an error when we already got the courses but somehow failed to get the groups.